Improve shader function descriptions

AshbyGeek opened the pull request:
Adds much detail to the documentation on built-in shader functions, copied from (with a few modifications) and linked to official GLSL documentation.

That documentation was moved into a separate file due to its verbosity, and organized into smaller sections following the organizational comments in https://github.com/godotengine/godot/blob/master/servers/rendering/shader_language.cpp.

Style tries to generally follow the styling used in class documentation, but differences between .gdshader and .gdscript meant some changes had to be made.

_Bugsquad edit:_ Closes https://github.com/godotengine/godot-docs/issues/5594 and closes https://github.com/godotengine/godot-docs/issues/9310, supersedes https://github.com/godotengine/godot-docs/pull/5729.

AshbyGeek commented:
Pull request for issue #9310

skyace65 commented:
CC @clayjohn

clayjohn commented:
That was unbelievably fast. I'm going to need a few days to review this. But I am very excited!

AThousandShips reviewed (commented):
This looks good, gonna go through it for style again but this was some that stuck out for now

AThousandShips commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,5273 @@
+
+.. note::
+    Most operations on vectors (multiplication, division, etc) are performed component-wise.
+    For example, the operation ``vec2(3,4) * vec2(10,20)`` would result in ``vec2(3 * 10, 4 * 20)``.
```
```suggestion
    For example, the operation ``vec2(3, 4) * vec2(10, 20)`` would result in ``vec2(3 * 10, 4 * 20)``.
```

Chaosus commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,5273 @@
+    +-----------------+--------------------------------------------------+--------------------------+
+    | vec_uint_type   | uint, uvec2, uvec3, or uvec4                     | genUType                 |
+    +-----------------+--------------------------------------------------+--------------------------+
+    | vec_bool_type   | bool, bvec2, bvec3, or bvec4                     | genUType                 |
```
```suggestion
    | vec_bool_type   | bool, bvec2, bvec3, or bvec4                     | genBType                 |
```
typo

Chaosus commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,5273 @@
+
+.. rst-class:: classref-method
+
+ivec2 **textureSize** ( samplerCubeArray s, int lod )
```
The following text is just a duplicated information, I think you should remove it, and instead add the overloads to header of first declaration eg:

```
ivec2 **textureSize** ( |gsampler2D| s, int lod )
ivec2 **textureSize** ( samplerCube s, int lod )
ivec3 **textureSize** ( |gsampler2DArray| s, int lod )
ivec3 **textureSize** ( |gsampler3D| s, int lod )
```

And do the same for other functions...

AshbyGeek replied:
There were a number of these kinds of functions, in a few of them the text was exactly the same, but a lot of them had minor differences. For instance, the `atan` function has significant behavioral differences between its two overloads. 

I decided that the documentation would be slightly clearer if these were addressed individually and that consistency of layout/format (in this case meaning each overload gets its own documentation) was more important than concerns about repetition.

Admittedly, most of them are not-just the difference between int and uint or similarly small, but even some of these behave differently, such as `packUnorm2x16` which has a different snippet of equivalent code for each overload.

AshbyGeek replied:
@Chaosus thoughts?

clayjohn replied:
I think we can separate them in cases like atan where you need two different descriptions. But in cases like `textureSize`, it makes sense to group them together. You are right not to worry about repetition, but some cases (like all the `texture*` functions) the sheer amount of repetition makes the document harder to navigate

AshbyGeek replied:
Understood, I will add this to my todo list.

clayjohn commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,5273 @@
+
+vec_type atan( |vec_type| y_over_x)
+
+    Calculate the arctangent given a tangent value of ``y/x``. Note: becuase of
```
```suggestion
    Calculate the arctangent given a tangent value of ``y/x``. Note: because of
```

clayjohn commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,5273 @@
+
+vec_type atanh( |vec_type| x)
+
+    Calculate the arctangent given a tangent value of ``y/x``. Note: becuase of
```
```suggestion
    Calculate the arctangent given a tangent value of ``y/x``. Note: because of
```

clayjohn commented on tutorials/shaders/shader_reference/shading_language.rst:
```diff
@@ -1122,371 +1122,15 @@ When using per-instance uniforms, there are some restrictions you should be awar
 - :ref:`Sky shaders <doc_sky_shader>`
 - :ref:`Fog shaders <doc_fog_shader>`
-
-Built-in functions
-------------------
-
-A large number of built-in functions are supported, conforming to GLSL ES 3.0.
-When vec_type (float), vec_int_type, vec_uint_type, vec_bool_type nomenclature
-is used, it can be scalar or vector.
```
I think its worth keeping this section and only having one sentence that simply points to the new page. Users may have a hard time finding it otherwise

clayjohn commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,5273 @@
+
+    If  any of these are specified for multiple parameters, they must all be the same type unless otherwise noted.
+
+.. note::
+    Most operations on vectors (multiplication, division, etc) are performed component-wise.
+    For example, the operation ``vec2(3, 4) * vec2(10, 20)`` would result in ``vec2(3 * 10, 4 * 20)``.
+    or the operation ``min(vec2(3, 4), vec2(1, 8))`` would result in ``vec2(min(3, 1), min(4, 8))``.
+
+    The `GLSL Language Specification <http://www.opengl.org/registry/doc/GLSLangSpec.4.30.6.pdf>` says under section 5.10 Vector and Matrix Operations:
+
+        With a few exceptions, operations are component-wise. Usually, when an operator operates on a
+        vector or matrix, it is operating independently on each component of the vector or matrix,
+        in a component-wise fashion. [...] The exceptions are matrix multiplied by vector,
+        vector multiplied by matrix, and matrix multiplied by matrix. These do not operate component-wise,
+        but rather perform the correct linear algebraic multiply.
```
This is probably better suited in the section on data types rather than here.

https://docs.godotengine.org/en/latest/tutorials/shaders/shader_reference/shading_language.html#data-types

AshbyGeek replied:
For now I copied the note into the main shading language page as well. If that works for you: great. If not, let me know what you would like me to try next.

clayjohn replied:
I think what makes this awkward is the entire section is talking about _operations_ not _functions_. GLSL isn't a language where you can override operators or customize them in any way. They are, for all intents and purposes, parts of the data type, not custom functions. 

But, reflecting on your comments, perhaps the best spot is in the section dedicated to operators https://docs.godotengine.org/en/latest/tutorials/shaders/shader_reference/shading_language.html#operators

AshbyGeek replied:
Ah, yes I can see how that would be a little offsetting. Putting it in the operators section is a good idea.

However, as far as I can tell this principle also applies to all the built in functions that have 2 vector parameters. This is why I believe that some version of this note still belongs in the built-in-functions section. I have three thoughts on ways to make this more clear:

1. I could change the wording of the existing note. Perhaps change all instances of the word 'operator' to 'function': "most functions that accept 2 vector parameters work component-wise..." Could also shorten the note a bit by not mentioning operators at all.
2. Remove the existing note and make a shorter one that is applied to each function that accepts two vector parameters.  Since we have a lot of overloads that are listed together, might have to call out specific method signatures in the note, but this would beore explicitly clear than a single generic note at the top.
3. Remove this note and trust that people writing shaders will learn this principle elsewhere or otherwise figure it out on their own. Moving the note to other pages more or less counts too, since this applies to many functions on this page and they may not think to look for such notes on other pages. Though, I agree that it is also applicable on other pages.
4. Move the content of this note to a new "component-wise operations" header on the shading language page and then add very brief notes ("component-wise function") that hyperlink to that section wherever they are applicable.

AshbyGeek replied:
@clayjohn Thoughts?

tetrapod00 replied:
My initial thought would be to add a single adjective "per-component" or "component-wise" to any relevant functions.  I suppose you could even put "Component-wise." as a whole sentence if the adjective doesn't fit naturally. Probably still keep a section somewhere else that has a more in-depth explanation. This accomplishes the goal of minimal text, while having each function be more or less self-documenting, *even if linked to on its own, to a new shader user*.

Which functions should it be added to? Either:
- Any functions that take in a `vec_type` parameter and does per-component operations. The problem is there are a *lot* of these.
- Any functions that take in *multiple* `vec_type` parameters and does per-component operations. Maybe the single parameter functions obviously work component-wise, since they take a vec4 and return a vec4.
---
I also like option 1. I would mention "operators and functions", and use an example of a function in addition to an operation, maybe something simple like `abs()` or `max()`.
I like the idea of option 4, in conjunction with a "component-wise" as a single word link in each function description, if that doesn't bloat the links too much.

AshbyGeek replied:
> Which functions should it be added to? Either:
> 
>  - Any functions that take in a vec_type parameter and does per-component operations. The problem is there are a lot of these.
>  - Any functions that take in multiple vec_type parameters and does per-component operations. Maybe the single parameter functions obviously work component-wise, since they take a vec4 and return a vec4.

Yeah, there's more than 100 functions listed here, almost all of them deal with vectors. I don't really have time right now to confirm which ones operate componentwise (almost all of them) and which ones actually perform an algebraic function. So I'm going to assume that its obvious that a single parameter function works component-wise.

AshbyGeek replied:
Also, I'm working on it now.

AshbyGeek replied:
Well, I discovered that the official [OpenGL ES 4.0 specification](https://registry.khronos.org/OpenGL/specs/gl/GLSLangSpec.4.30.pdf) is very good about specifying which functions perform component-wise. So I was actually able to just put a note on each function that does so. Will have a new commit up shortly.

AshbyGeek replied:
ok, @tetrapod00  and @clayjohn, just pushed up a commit that puts a Hyperlink above the description of each relevant function.

tetrapod00 replied:
I like this solution.
For entries that have a standalone "componentwise" link, I'd put the link after the description and before the parameter table, so the more relevant info about what the function actually does comes first.

I would wait until clayjohn approves this solution in general before making any more tedious changes, though.

clayjohn reviewed (commented):
Amazing work! This looks really good so far. 

I think I just have two general comments:
1. I agree with Chaosus' comment earlier that, in most cases, functions with the same name should be grouped instead of duplicated. Especially for functions with really long descriptions. 
2. I like how you kept the short description in the table. But I think it needs to be truly a short description. Some of the entries have the entire description copied in the table and that ends up making the table bloated in hard to navigate. 

See DfDx for example
![image](https://github.com/godotengine/godot-docs/assets/16521339/7fc2e260-5050-4b64-ab3c-80aa8ff14cac)

In these cases, the table actually has more information than the long description below.

clayjohn commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,5273 @@
+
+    .. note::
+
+        Available only in the fragment shader
```
```suggestion
        Available only in the fragment shader
        Not available when using the GL_Compatibility rendering backend.
```

The same note should be added for dFdyCoarse, dFdxFine, dFdyFine, fwidthFine, and fwidthCoarse

AshbyGeek commented:
Awesome! I'm glad you like it.  I'm pushing hard to finish a project milestone this week, but I should be able to work on an update next week.

skyace65 commented:
@AshbyGeek is this something you're still able to work on?

AshbyGeek commented:
Yes! I'm still able to work on it, just completely forgot about it for a minute. It will probably be Thursday before I have time to do it though. I just set a calendar reminder to help me keep track of it this time.

AshbyGeek commented:
Ok, that should get this branch up to date with all the above comments. Render for your next round of reviews!

skyace65 reviewed (approved):
Did a grammar and spelling check and everything looks good.

tetrapod00 commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,3781 @@
+    Return the sine of the parameter.
+
+    :param angle:
+        takehe quantity, in radians, of which to return the sine
```
Typo?

AshbyGeek replied:
```suggestion
        the quantity, in radians, of which to return the sine
```

tetrapod00 commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,3781 @@
+
+|vec_int_type| **bitfieldExtract** ( |vec_int_type| value, int offset, int bits )
+
+    Eextracts a subset of the bits of value and returns it in the least significant bits of the result.
```
```suggestion
    Extracts a subset of the bits of value and returns it in the least significant bits of the result.
```
Typo

tetrapod00 commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,3781 @@
+
+|vec_type| **roundEven** ( |vec_type| x )
+
+    returns a value equal to the nearest integer to x.
```
```suggestion
    Returns a value equal to the nearest integer to x.
```
Capitalization

tetrapod00 commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,3781 @@
+
+|vec_uint_type| **bitfieldExtract** ( |vec_uint_type| value, int offset, int bits )
+
+    Eextracts a subset of the bits of value and returns it in the least significant bits of the result.
```
```suggestion
    Extracts a subset of the bits of value and returns it in the least significant bits of the result.
```
Typo

tetrapod00 commented:
I also noticed no consistency of "Return" vs "Returns" or "Calculate" vs "Calculates" in the descriptions of functions, but I don't think that's really worth the effort to standardize.

AshbyGeek commented:
@tetrapod00 Any thoughts on the conversation @clayjohn and I were having about [functions working on arrays piecewise](https://github.com/godotengine/godot-docs/pull/9338#discussion_r1609080886)?

AshbyGeek commented:
Latest commit also improves on some formatting and fixes some oddities with the table of contents on the left. 

- made the classref-method styling apply to entries with multiple overloads 
- used a single table row for entries with multiple overloads, but figured out how to keep the return types and the rest 
- 2nd column of the tables don't allow text wrapping (had to add to static/custom.css) so that table rows with multiple entries will keep those entries aligned
- Fixed some function names in the descriptions that weren't bold
- Fixed a couple entries that failed to include the on-hover description of the vec_type

AThousandShips commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,3804 @@
+    +-----------------------------------+-----------------------------------------------------+
+    | Operation                         | Equivalent Scalar Operation                         |
+    +===================================+=====================================================+
+    | ``sqrt(vec2(4,64))``              | ``vec2(sqrt(4),sqrt(64))``                          |
```
```suggestion
    | ``sqrt(vec2(4, 64))``             | ``vec2(sqrt(4), sqrt(64))``                         |
```
For all these

tetrapod00 replied:
```suggestion
    | ``sqrt(vec2(4.0, 64.0))``              | ``vec2(sqrt(4.0), sqrt(64.0))``                          |
```
Across all the shader pages, floats are styled with the explicit decimal to distinguish from ints.

AThousandShips commented on tutorials/shaders/shader_reference/shading_language.rst:
```diff
@@ -469,33 +469,59 @@ Operators
+    +--------------------------------------+-----------------------------------------------------+
+    | Operation                            | Equivalent Scalar Operation                         |
+    +======================================+=====================================================+
+    | ``vec3(4,5,6) + 2``                  | ``vec3(4+2, 5+2, 6+2)``                             |
```
```suggestion
    | ``vec3(4, 5, 6) + 2``                | ``vec3(4 + 2, 5 + 2, 6 + 2)``                       |
```

AThousandShips commented on tutorials/shaders/shader_reference/shading_language.rst:
```diff
@@ -469,33 +469,59 @@ Operators
+    +--------------------------------------+-----------------------------------------------------+
+    | ``vec2(3, 4) * vec2(10, 20)``        | ``vec2(3 * 10, 4 * 20)``                            |
+    +--------------------------------------+-----------------------------------------------------+
+    | ``mat2(vec2(1,2),vec2(3,4)) + 10``   | ``mat2(vec2(1+10,2+10),vec2(3+10,4+10))``           |
```
```suggestion
    | ``mat2(vec2(1, 2), vec2(3, 4)) + 10``| ``mat2(vec2(1 + 10, 2 + 10), vec2(3 + 10,4 + 10))`` |
```

AThousandShips reviewed (approved):
Otherwise I think this looks good

tetrapod00 commented:
I see a lot of places where style can be improved in a minor way, like styling literals as code and with a decimal marker for floats (e.g. `0.0` instead of `0` or "0.0" for a float value of 0). I'll do a detail-oriented review on this (or a followup PR if we'd rather not block merging).

tetrapod00 commented on tutorials/shaders/shader_reference/shader_functions.rst:
```diff
@@ -0,0 +1,3804 @@
+==================
+
+Godot supports a large number of built-in functions, conforming roughly to the
+GLSL 4.00 specification.
```
The [shading language](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/shading_language.html#introduction) page says godot's shading language is based on `GLSL ES 3.0`. So one of these two pages is wrong. (Or for built-in functions, there is no difference between 3.0 and 4.0)

AshbyGeek replied:
I need to go some digging into Godot's source to try to find the answer because I saw that too, but some of these functions weren't listed in the opengl es 3.0 documentation I was originally looking in. So need to find a definitive source of truth for which version is supported.

tetrapod00 replied:
If you blame the shading language page, that phrasing was added in [this commit ](https://github.com/godotengine/godot-docs/commit/ff696205cea64ca6b624c918c53fdd90ba4e2482). It's only had minor phrasing and formatting changes since then.

AshbyGeek replied:
Oh, lol. Committed 7 years ago. Makes me think very strongly that the info is just outdated.

Now I'm a little worried that the docs and the code have diverged a bit. Maybe I should compare the functions listed in docs to those actually supported by code.

tetrapod00 replied:
An easy solution would be to only specify "based on GLSL" without the version and link to the current version of the GLSL spec.
Ideally we would coordinate the change in wording across this page, the Shading Language page, and the migrating from GLSL page.
For this PR I think you just need to make sure that you're not *adding* wrong information, the other cases can be changed separately.

AshbyGeek replied:
Well, trolled the code and other things for a while before I realized that I was probably confusing OpenGL ES shader language (commonly ESSL or GLSL ES) and OpenGL Shader Language (GLSL). Turns out there isn't an ESSL 4.5, so I must have looked at GLSL documentation .

Wikipedia says that the two languages are related, but not compatible, so that's fun. So yeah, there's quite a few references in the engine code to GLES3, though there's also a lot of glsl files that reference "#version 450" which is definitely a GLSL version. So there seems to be substantial differences in shader compilation for Vulkan vs. D3D12 vs Legacy OpenGL backends. Guess that's why Godot needs its own shader language.

So yeah, I definitely need to change that statement.

AshbyGeek replied:
Ok, pushed a commit that fixed that wording, and the couple formatting issues you specifically mentioned above. I'll probably leave the rest of the formatting issues for you.

tetrapod00 commented:
I started on a nitpicky style review but I think it would be better as a followup PR - making all the required changes is annoying as suggestions.
It's a lot of stuff like:
"if `x < -1`" vs "if x is less than 0.0" vs "if x ≤ 0"
``(e^x + e^-x)/2`` vs "x=2\ :sup:`y`"
This stuff should be standardized, and I can do the work, but it will be annoying to do as a bunch of individual suggestions.

IMO the style inconsistency is not bad enough to prevent merging as is.

AshbyGeek commented:
Ok, fixed the couple things you already mentioned since I needed to fix my GLSL 4.0 mistake. I'm more than happy to let you fix the rest of the tedious formatting issues in another PR. 

What does the merge process look like from here? We wait for ClayJohn or some such to come give a final approval and pull it in?

tetrapod00 commented:
Yeah, now you're waiting on another review from clayjohn or someone else with shader-specific knowledge to do a high-level review including whatever other style nitpicks come up, then it should be merge-able. IMO it's already so much better than it was that it is currently merge-able as-is. 

(Also I believe in most cases we would address the style nitpicks in this PR, not a followup, but it is really tedious work and IMO not blocking.)

clayjohn commented:
I keep meaning to do a deep review of this. But this is a huge chunk of work and I never have the time to do it in one sitting. 

At this point I trust that the current state is way better than it was before. I also trust that the reviews from tetrapod and ATS would catch any glaring issues. Further Ashby has shown that they are working hard to make this as good as possible and I appreciate and respect  his efforts. 

I suggest we go ahead with merging this without me doing a super detailed review as I don't want to be the blocker for this any longer and I trust all the others involved in the process. 

Great work everyone!

tetrapod00 commented:
I did an in-depth review. Overall this looks great. I found a few factual and copy-paste errors, and made a lot of minor style changes. 

The changes are in a PR to the branch at https://github.com/AshbyGeek/godot-docs/pull/1, which seemed to be the best way to make changes this big. Summary of changes:
- Standardized style.
- Fix a few factual and copy paste errors.
- Fix some formatting.
- Consistently use code blocks for literals.
- Use sentence case for all parameters.
- Avoid math symbols, prefer code-style `>=`.

AshbyGeek commented:
Wowee, you weren't kidding about a lot of fiddly changes. You have a much better eye for detail and consistency in technical documentation than I do. My hat is off to you sir! I have already merged that pull request.

My only real question from looking at your pull request is really just a code style question:

You changed a bunch of instances of 
```
.. rst-class:: classref-method

| |vec_type| **abs**\ (\ |vec_type| x)
| |vec_int_type| **abs**\ (\ |vec_int_type| x)
```
to the much more verbose
```
.. rst-class:: classref-method

|vec_type| **abs**\ (\ |vec_type| x)

.. rst-class:: classref-method

|vec_int_type| **abs**\ (\ |vec_int_type| x)

```

As far as I can tell the rendered HTML is the same. Am I mistaken?

Note, as I said above I've already merged your pull request. If you want to change these back, you can send me another pull request I guess, lol.

tetrapod00 commented:
The rendered HTML is not the same (tested on Windows 10, Firefox, 100% browser zoom). Before the changes, the function names are gray, but after the changes the function names are white, like the functions which appear on a single line with no overloads:
Before the changes, in https://github.com/godotengine/godot-docs/pull/9338/commits/220fb2b85988899f5b2676bc61b6b1e5ba49efd6:
![firefox_KJ2wTAiJDT](https://github.com/user-attachments/assets/0be39ffa-5420-484d-8d50-a49aa866fbf4)

After the changes, in https://github.com/godotengine/godot-docs/pull/9338/commits/8972c45213e225ecf0048a183998610575257aa0:
![firefox_1WI53dZu1b](https://github.com/user-attachments/assets/7eff65cc-b678-43cd-9449-b6d9ec6c463f)

Compare to any function without overloads:
![firefox_nmIoNbaDA4](https://github.com/user-attachments/assets/9a483f7f-f62d-40b7-8427-d4f5982e64db)

The version with white function names is correct.

(In https://github.com/godotengine/godot-docs/pull/9338/commits/220fb2b85988899f5b2676bc61b6b1e5ba49efd6, there were also a few unrelated formatting errors like the following, but those are not related to this difference.)
![image](https://github.com/user-attachments/assets/da80165c-2a5a-4528-9485-84187a1de74f)
![image](https://github.com/user-attachments/assets/d58a0c0e-a812-405b-b0f9-d840d4ee4cbf)

tetrapod00 reviewed (approved):
With the style changes applied, this is good and mergable.

There are still many minor improvements that *could* be made, and I fully expect to find a factual error or two even after this is merged. But this is already much better than the current state of the shader function documentation. It's a good base for future improvements, which can be done in smaller chunks.

AshbyGeek commented:
Why does yours render dark? Godot docs online use a light theme, which is also what I see on my computer.

Here's the code from commit 220fb2b:
```
.. rst-class:: classref-method

| |gvec4_type| **textureProjGrad**\ (\ |gsampler2D| s, vec3 p, vec2 dPdx, vec2 dPdy)
| |gvec4_type| **textureProjGrad**\ (\ |gsampler2D| s, vec4 p, vec2 dPdx, vec2 dPdy)
| |gvec4_type| **textureProjGrad**\ (\ |gsampler3D| s, vec4 p, vec3 dPdx, vec3 dPdy)
```

when compiled with Sphinx produces the following HTML on my computer:
```
<div class="classref-method line-block" id="shader-func-textureprojgrad">
    <div class="line">
        <abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr>
        <strong>textureProjGrad</strong>
        (<abbr title="Any of: sampler2D, isampler2D, uSampler2D">gsampler2D</abbr>
        s, vec3 p, vec2 dPdx, vec2 dPdy)
    </div>
    <div class="line">
        <abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr>
        <strong>textureProjGrad</strong>
        (<abbr title="Any of: sampler2D, isampler2D, uSampler2D">gsampler2D</abbr>
        s, vec4 p, vec2 dPdx, vec2 dPdy)
    </div>
    <div class="line">
        <abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr>
        <strong>textureProjGrad</strong>
        (<abbr title="Any of: sampler3D, isampler3D, uSampler3D">gsampler3D</abbr>
        s, vec4 p, vec3 dPdx, vec3 dPdy)
    </div>
</div>
```
which looks like this in Chrome v129.0.6668.101 (on Windows 10)
![image](https://github.com/user-attachments/assets/2d6ae265-e36c-4288-88e1-fb5f0c485592)

----------------------------------------------

Here's code from commit 897nc452 (the merge commit) 
```
.. _shader_func_textureProjGrad:

.. rst-class:: classref-method

|gvec4_type| **textureProjGrad**\ (\ |gsampler2D| s, vec3 p, vec2 dPdx, vec2 dPdy)

.. rst-class:: classref-method

|gvec4_type| **textureProjGrad**\ (\ |gsampler2D| s, vec4 p, vec2 dPdx, vec2 dPdy)

.. rst-class:: classref-method

|gvec4_type| **textureProjGrad**\ (\ |gsampler3D| s, vec4 p, vec3 dPdx, vec3 dPdy)
```

which produces the following html

```
<p class="classref-method" id="shader-func-textureprojgrad">
    <abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr>
    <strong>textureProjGrad</strong>
    (<abbr title="Any of: sampler2D, isampler2D, uSampler2D">gsampler2D</abbr>
    s, vec3 p, vec2 dPdx, vec2 dPdy)
</p>
<p class="classref-method">
    <abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr>
    <strong>textureProjGrad</strong>
    (<abbr title="Any of: sampler2D, isampler2D, uSampler2D">gsampler2D</abbr>
    s, vec4 p, vec2 dPdx, vec2 dPdy)
</p>
<p class="classref-method">
    <abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr>
    <strong>textureProjGrad</strong>
    (<abbr title="Any of: sampler3D, isampler3D, uSampler3D">gsampler3D</abbr>
    s, vec4 p, vec3 dPdx, vec3 dPdy)
</p>
```

and looks like this in Chrome v129.0.6668.101 (on Windows 10)
![image](https://github.com/user-attachments/assets/fb9bad25-3d06-461a-b36f-79c099b8bf2b)

-----------------------------------------------------------

I would guess from the inclusion of the "|" in your last image and the fact that you have a dark theme, that there's some configuration difference between your sphinx compiler and mine or maybe you've set a custom theme for your Godot docs? 🤷 

As I've said, the code change here really isn't a show stopper for me, I just find my previous syntax easier to read.

tetrapod00 commented:
The docs was dark mode due to my Firefox configuration. I switched Firefox to light mode.

Now this reStructuredText source from commit https://github.com/godotengine/godot-docs/commit/220fb2b85988899f5b2676bc61b6b1e5ba49efd6:
```reStructuredText
| |gvec4_type| **textureProjGrad**\ (\ |gsampler2D| s, vec3 p, vec2 dPdx, vec2 dPdy)
| |gvec4_type| **textureProjGrad**\ (\ |gsampler2D| s, vec4 p, vec2 dPdx, vec2 dPdy)
| |gvec4_type| **textureProjGrad**\ (\ |gsampler3D| s, vec4 p, vec3 dPdx, vec3 dPdy)
```
 Produces this HTML:
```html
<div class="classref-method line-block" id="shader-func-textureprojgrad">
<div class="line"><abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr> <strong>textureProjGrad</strong>(<abbr title="Any of: sampler2D, isampler2D, uSampler2D">gsampler2D</abbr> s, vec3 p, vec2 dPdx, vec2 dPdy)</div>
<div class="line"><abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr> <strong>textureProjGrad</strong>(<abbr title="Any of: sampler2D, isampler2D, uSampler2D">gsampler2D</abbr> s, vec4 p, vec2 dPdx, vec2 dPdy)</div>
<div class="line"><abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr> <strong>textureProjGrad</strong>(<abbr title="Any of: sampler3D, isampler3D, uSampler3D">gsampler3D</abbr> s, vec4 p, vec3 dPdx, vec3 dPdy)</div>
</div>
```
Formatted similar to yours, it is identical:
```html
<div class="classref-method line-block" id="shader-func-textureprojgrad">
    <div class="line">
        <abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr>
        <strong>textureProjGrad</strong>
        (<abbr title="Any of: sampler2D, isampler2D, uSampler2D">gsampler2D</abbr>
        s, vec3 p, vec2 dPdx, vec2 dPdy)
    </div>
    <div class="line">
        <abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr>
        <strong>textureProjGrad</strong>
        (<abbr title="Any of: sampler2D, isampler2D, uSampler2D">gsampler2D</abbr>
        s, vec4 p, vec2 dPdx, vec2 dPdy)
    </div>
    <div class="line">
        <abbr title="Any of: vec4, ivec4, uvec4">gvec4_type</abbr>
        <strong>textureProjGrad</strong>
        (<abbr title="Any of: sampler3D, isampler3D, uSampler3D">gsampler3D</abbr>
        s, vec4 p, vec3 dPdx, vec3 dPdy)
    </div>
</div>
```

Which looks like this in `Firefox 131.0.3 (64-bit)`, Windows 10:

![firefox_WddBOo7TJf](https://github.com/user-attachments/assets/fc4dd8e9-0f79-427d-b5ec-3259ca071ad2)

Now it looks *fine* but some style is still not applied, still. Compared to another function, the compact style still doesn't produce the gray text for the rest of the function.
![firefox_L8pscPHq0G](https://github.com/user-attachments/assets/953a3d00-c220-4e9e-bc7c-4fc9eb6045f6)

On `Edge Version 130.0.2849.46`, it's the same:
![msedge_LlkxoWPgtf](https://github.com/user-attachments/assets/879cb718-8c13-48d7-a18a-1d80b65fa5ca)
![msedge_ZyqhzMRei0](https://github.com/user-attachments/assets/3b0c2d1b-28e2-4179-905c-adffd116c569)

On `Chrome Version 130.0.6723.70`, I can't find a way to force light mode (the main docs site is also forced dark mode). But the difference in font style persists:
![chrome_XMHFLkLXGT](https://github.com/user-attachments/assets/c6ee71c6-2eb7-4040-8c11-c41ad1f03fd8)
![chrome_WSUY3QJjrq](https://github.com/user-attachments/assets/2e651116-3668-4a63-ab8a-22a760074c93)

Since the HTML is identical, this is certainly some difference in configuration between your computer and mine. The more compact style (from https://github.com/godotengine/godot-docs/commit/220fb2b85988899f5b2676bc61b6b1e5ba49efd6) is probably the better choice in the abstract. The more verbose style (from https://github.com/godotengine/godot-docs/pull/9338/commits/8972c45213e225ecf0048a183998610575257aa0) does seem to render more consistently, with this tiny amount of testing.

Note that if we switch back to the compact style, there *are* some unrelated formatting errors that were in (from https://github.com/godotengine/godot-docs/commit/220fb2b85988899f5b2676bc61b6b1e5ba49efd6. The stray `|` is one of those formatting errors. It's not directly related to the question of which style to use.

With just the two of us testing, I'm inclined to keep the current verbose style and investigate again after this is merged.

AshbyGeek commented:
Sounds like a plan to me. And honestly, don't even feel like additional investigation is really necessary. More consistent/better results should be a higher priority than code style.

tetrapod00 commented:
This closes https://github.com/godotengine/godot-docs/issues/5594.

tetrapod00 commented:
And, finally merged. Thanks for this *monumental* improvement to the shader documentation, and thanks for your consistent work in response to reviews. For a first-time contribution, this is huge.

And thanks for all the reviews, everyone!

AshbyGeek commented:
Woohoo! Glad I could help!
